### PR TITLE
Integration

### PR DESCRIPTION
**Pull request by vytautassugintas**
Please check our tests and give us some tips what to test. Also, it is very important to us to know how to mock HTTP request.

Thanks! Acorns to you

### CODE REVIEW

File: src/app/acorns/acorns.controller.js [hunk 1]
```diff
@@ -5,56 +5,53 @@
  * Created by vytautassugintas on 06/04/16.
```
Comment (Medds) [inline src/app/acorns/acorns.controller.js]: Remove unnecessary comments. Especially duplicates

Comment (Medds) [inline src/app/acorns/acorns.controller.js]: Remo unnecessary comments. Especially duplicates

File: src/app/acorns/acorns.controller.js [hunk 2]
```diff
@@ -5,56 +5,53 @@
...
+            function checkUser() {
+                ProfileService.checkUser().then(function (val) {
+                    val.logged ? $window.location.href = "#/acorns" : $window.location.href = "#/login";
+                });
+            }
+        };
+
+        AcornController.$inject = ['$http', '$scope', '$window', '$cookies', '$timeout', '$httpParamSerializer', '$filter', 'ProfileService', 'Challenges', 'Resources'];
+
+        angular
```
Comment (Medds) [inline src/app/acorns/acorns.controller.js]: Rule 1 of the SOLID principles and Rule 1 of JohnPapas  styleguide - Single Responsibility principle. Better to define one component per file. https://github.com/johnpapa/angular-styleguide/blob/master/a1/README.md#style-y001

Here, you're declaring both the module and the controller in the same file, which might make it harder to find the module. Probably easier to move it out to `acorn.module.js` or something similar, so it's easily findable.

File: src/app/acorns/acorns.controller.js [hunk 3]
```diff
@@ -5,56 +5,53 @@
...
-        function isValid(value) {
-            return typeof value === "undefined";
-        }
-
-        function checkUser() {
-            ProfileService.checkUser().then(function (val) {
-                val.logged ? $window.location.href = "#/acorns" : $window.location.href = "#/login";
+        'use strict';
+        var AcornController = function ($http, $scope, $window, $cookies, $timeout, $httpParamSerializer, $filter, ProfileService, Challenges, Resources) {
+            $scope.showLoader = true;
```
Comment (Medds) [inline src/app/acorns/acorns.controller.js]: Rule 7 in JohnPapas Styleguide - https://github.com/johnpapa/angular-styleguide/blob/master/a1/README.md#style-y030

Try to use controller as, instead of $scope as much as possible. ControllerAs makes scope pollution much less of a problem and helps you nest, access and control lots of scopes.

File: src/app/app.module.js [hunk 4]
```diff
@@ -15,8 +15,8 @@
             'myApp.components.notifications',
```
Comment (Medds) [inline src/app/app.module.js]: Modules are meant to represent a logical grouping to the angular application, where some parts of the application might be self-contained enough to be used individually in other applications. Over here, components are pretty coupled, which means that they could just be used under the same module, which would remove much of the need for having different modules for every component and then spamming them all in this module declaration.

File: src/app/app.utils.js [hunk 5]
```diff
@@ -17,4 +17,28 @@ function acornPlural(amount) {
```
Comment (Medds) [inline src/app/app.utils.js]: You should never pollute the global scope with your own functions, unless you're making a helper library, where you would define an access variable. Better idea is to have these helper utils in an `angular.factory` and inject them where necessary. This makes it easier to test them and provides clear separation.

File: src/app/components/kudos-add-new-idea/kudos-add-new-idea.component.js [hunk 6]
```diff
@@ -37,10 +39,10 @@
                 showAddIdeaFormErrorMessage("Please enter idea");
                 return false;
             } else if ($scope.author.length > 30) {
```
Comment (Medds) [inline src/app/components/kudos-add-new-idea/kudos-add-new-idea.component.js]: Why not just have `maxlength` attributes on the HTML of the input? This would remove the need for any if this logic in the controller

File: src/app/components/kudos-challenge-completed/kudos-challenge-completed.service.js [hunk 7]
```diff
@@ -0,0 +1,20 @@
+(function () {
+    "use strict";
+    angular.module("myApp.components.completedChallenges")
+        .factory("CompletedChallengesService", CompletedChallengesService);
+
+    CompletedChallengesService.$inject = [
+        "Challenges"
+    ];
+
+    function CompletedChallengesService(challengesBackend) {
```
Comment (Medds) [inline src/app/components/kudos-challenge-completed/kudos-challenge-completed.service.js]: It's only my opinion - but is this service actually necessary? You're wrapping an `angular.factory` inside another `angular.factory` just for the sake of it. Usually, services have some sort of logical purpose, where they contain several service methods which belong to the same logical area and provide different parts of info for the same view. This service just pipes the return response of one  method from the backend- doesn't even return a promise or anything.  IMO, this could be combined with `KudosChallengeNewService` and, probably, something else to make a unified factory which deals with providing challenge info in a logical manner.

File: src/app/components/kudos-challenge-completed/kudos-challenge-completed.service.js [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+        var service = {
+            completedChallenges: CompletedChallenges
+        };
+        return service;
+
+        function CompletedChallenges() {
+            return challengesBackend.completedChallenges();
```
Comment (Medds) [inline src/app/components/kudos-challenge-completed/kudos-challenge-completed.service.js]: The naming of the method doesn't say anything. Function/method names should be self explanatory and usually begin with a verb, so this method is probably a `getCompletedChallenges`? `fetchCompletedChallenges`. A developer should know what the method does just by looking at the name. As Robert C. Martin(Uncle Bob - creator of the "Clean Code" series) said - function and variables names are the best documentation.

File: src/app/acorns/acorns.controller.js [hunk 9]
```diff
@@ -5,56 +5,53 @@
...
+
+            $scope.isValid = isValid;
+            $scope.checkUser = checkUser;
+
+            checkUser();
```
Comment (Medds) [inline src/app/acorns/acorns.controller.js]: It's common to have one `activate()` function in a controller/directive where all the initialization logic is contained, so other developers know where everything is initialized. `angular.component` even has an `$onInit` lifecycle hook exactly for that reason. It's probably a good idea to move this into the `activate()` function so all the init logic is located in one place.

File: src/app/acorns/acorns.html [hunk 10]
```diff
@@ -3,7 +3,7 @@
 <kudos-navbar></kudos-navbar>
 
 <div class="container">
-    <div class="panel panel-body">
+   <!-- <div class="panel panel-body">
```
Comment (Medds) [inline src/app/acorns/acorns.html]: There should never be commented code in releasable code, be it javascript, css or html

File: src/app/components/kudos-add-new-idea/kudos-add-new-idea.component.js [hunk 11]
```diff
@@ -37,10 +39,10 @@
                 showAddIdeaFormErrorMessage("Please enter idea");
```
Comment (Medds) [inline src/app/components/kudos-add-new-idea/kudos-add-new-idea.component.js]: Could a `placeholder` be used in the html here instead of an "error message"?

File: src/app/components/kudos-challenge-completed/kudos-challenge-completed.component.js [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+            showMoreButton(Resources.getCompletedChallenges());
+            $scope.showLessButton = false;
+        }
+
+    };
+
+    CompletedChallengesController.$inject = ['$scope', '$httpParamSerializer', 'CompletedChallengesService', 'Challenges', 'Resources'];
+
+    angular.module('myApp.components.completedChallenges', [])
+        .component('kudosChallengeCompleted', {
```
Comment (Medds) [inline src/app/components/kudos-challenge-completed/kudos-challenge-completed.component.js]: Single responsibility principle - one component per file. I'd have no problem if the controller was declared via function, but here you are declaring a new component.

File: src/app/components/kudos-challenge-completed/kudos-challenge-completed.component.js [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+                if (Resources.getCompletedChallenges()[index].participant == Resources.getCurrentUserEmail()) return "You won the challenge!"
+                else return Resources.getCompletedChallenges()[index].participant + " won the challenge";
+            } else if (Resources.getCompletedChallenges()[index].participantStatus == false){
+                if (Resources.getCompletedChallenges()[index].participant == Resources.getCurrentUserEmail()) return Resources.getCompletedChallenges()[index].creator + " won the challenge";
+                else return "You won the challenge!"
+            }
+        }
+
+        function showMoreButton(val) {
+            if (val.length > $scope.completedChallengesLimit) {
```
Comment (Medds) [inline src/app/components/kudos-challenge-completed/kudos-challenge-completed.component.js]: does 
```javascript
$scope.showMoreButton  = val.length > $scope.completedChallengesLimit
```
not produce the same result?

File: src/app/components/kudos-challenge-completed/kudos-challenge-completed.component.js [hunk 14]
```diff
@@ -0,0 +1,90 @@
...
+        $scope.acornPlural = acornPlural;
+        $scope.getWinner = getWinner;
+        $scope.userEmail = Resources.getCurrentUserEmail();
+        $scope.showMore = showMore;
+        $scope.showLess = showLess;
+        $scope.showMoreButton = showMoreButton;
+        var showMoreLimit = 5;
+        $scope.completedChallengesLimit = 5;
+
+        CompletedChallengesService.completedChallenges().then(function (val) {
```
Comment (Medds) [inline src/app/components/kudos-challenge-completed/kudos-challenge-completed.component.js]: Please look at my comment in this factory. Probably could use different naming and a different service altogether.

File: src/app/components/kudos-challenge-new/kudos-challenge-new.component.js [hunk 15]
```diff
@@ -0,0 +1,132 @@
...
+
+    KudosChallengeNewController.$inject = ['$httpParamSerializer', '$scope', 'KudosChallengeNewService', 'Challenges', 'Resources'];
+
+    angular.module('myApp.components.challengeNew', [])
+
+        .directive('kudosChallengeNew', function () {
+            return {
+                controller: 'KudosChallengeNewController',
+                restrict: 'E',
+                scope: false,
```
Comment (Medds) [inline src/app/components/kudos-challenge-new/kudos-challenge-new.component.js]: If you are creating a directive, though I'm certain this could be converted to `angular.component`, it's probably best to use an isolated scope, since you don't want scope pollution by accident and you already have a standalone controller for this. I've yet to find an example of a directive where inheriting the scope from a parent is the best solution of scope initiation.

File: src/app/components/kudos-challenge-new/kudos-challenge-new.component.js [hunk 16]
```diff
@@ -0,0 +1,132 @@
...
+                templateUrl: 'app/components/kudos-challenge-new/kudos-challenge-new.html'
+            }
+        })
+        .controller('KudosChallengeNewController', KudosChallengeNewController)
```
Comment (Medds) [inline src/app/components/kudos-challenge-new/kudos-challenge-new.component.js]: Single responsibility principle - one component per file. I'd have no problem if the controller was declared via function, but here you are declaring a new component.

File: src/app/components/kudos-challenge-new/kudos-challenge-new.html [hunk 17]
```diff
@@ -1,24 +1,46 @@
...
     <div ng-if="!id">
         <div class="alert alert-info acorn-bg margin-top">
             <span class="glyphicon glyphicon-info-sign"></span>
-            So far, no one Challenged you
+            You don't have any new challenges!
         </div>
     </div>
 </div>
 
+<!--
```
Comment (Medds) [inline src/app/components/kudos-challenge-new/kudos-challenge-new.html]: There should be no commented code in release.

File: src/app/components/kudos-challenge-new/kudos-challenge-new.component.js (hunk 15 above)
Comment (minijus) [inline src/app/components/kudos-challenge-new/kudos-challenge-new.component.js]: And 50 kudos goes to ... Jurgis!

Sick job mate! Beer from me :)

On Tue, Jun 28, 2016 at 11:04 PM Jurgis Bridžius <notifications@github.com>
wrote:

> In src/app/components/kudos-challenge-new/kudos-challenge-new.component.js
> <https://github.com/open-kudos/open-kudos-intern-web/pull/203#discussion_r68832162>
> :
>
> > +
> > +        function doesDateExist(index) {
> > +            return Resources.getNewChallenges()[index].finishDate == null;
> > +        }
> > +
> > +    };
> > +
> > +    KudosChallengeNewController.$inject = ['$httpParamSerializer', '$scope', 'KudosChallengeNewService', 'Challenges', 'Resources'];
> > +
> > +    angular.module('myApp.components.challengeNew', [])
> > +
> > +        .directive('kudosChallengeNew', function () {
> > +            return {
> > +                controller: 'KudosChallengeNewController',
> > +                restrict: 'E',
> > +                scope: false,
>
> If you are creating a directive, though I'm certain this could be
> converted to angular.component, it's probably best to use an isolated
> scope, since you don't want scope pollution by accident and you already
> have a standalone controller for this. I've yet to find an example of a
> directive where inheriting the scope from a parent is the best solution of
> scope initiation.
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/open-kudos/open-kudos-intern-web/pull/203/files/876ed2f2fa16af7dfbe117c81a8e322172940158#r68832162>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ADdxjYhgG6HN0ISXcNllfYLoLzUNZJAQks5qQX5egaJpZM4I82FT>
> .
>
-- 
Best regards,
Minijus Laukaitis

File: src/app/components/kudos-give-kudos-small/kudos-give-kudos-small.component.js [hunk 18]
```diff
@@ -14,11 +15,18 @@
             $scope.id = this.index;
         };
 
-        $scope.selectAutoText = function (text) {
-            $scope.sendKudosTo = text;
-            $scope.searchTermSelected = true;
-            $scope.autocompleteHide = true;
-        };
+        $scope.$watch(function () {
```
Comment (Medds) [inline src/app/components/kudos-give-kudos-small/kudos-give-kudos-small.component.js]: There seem to be very many `$scope.$watch`es around the place - is there no way to use more of a **tell, don't ask** pattern, where you don't rely on random variable changes so much? I know it would require changing some of the architecture, but `$scope.$watch` should be used as a last resort, since it isn't the most performant solution and makes the testing really awkward.

File: src/app/components/kudos-leaderboard-filter/kudos-leaderboard-filter.component.js [hunk 19]
```diff
@@ -0,0 +1,54 @@
...
+        }
+    };
+
+    LeaderboardFilterController.$inject = ['ProfileService', 'Resources', '$httpParamSerializer'];
+
+    angular.module('myApp.components.leaderboardFilter', [])
+        .component('kudosLeaderboardFilter', {
+            templateUrl: 'app/components/kudos-leaderboard-filter/kudos-leaderboard-filter.view.html',
+            controller: 'LeaderboardFilterController',
+            controllerAs: 'self'
```
Comment (Medds) [inline src/app/components/kudos-leaderboard-filter/kudos-leaderboard-filter.component.js]: An `angular.component` has `controllerAs` set to `$ctrl` by default without declaring anything. It's common practice to use this, since it leads to less code and less confusion. If you're not using the default accessor, you could set it to something semantically correct, which would be `filter` or `leaderboardFilter` for this instance, so it would be easily understandable.
TL;DR - best to not declare anything and use the default `$ctrl`, otherwise - semantify.

File: src/app/components/kudos-leaderboard-filter/kudos-leaderboard-filter.component.js [hunk 20]
```diff
@@ -0,0 +1,54 @@
...
+        })
+        .controller('LeaderboardFilterController', LeaderboardFilterController)
```
Comment (Medds) [inline src/app/components/kudos-leaderboard-filter/kudos-leaderboard-filter.component.js]: Single responsibility principle - one component per file. I'd have no problem if the controller was declared via function, but here you are declaring a new component.

File: src/app/components/kudos-user-history/kudos-user-history.component.js [hunk 21]
```diff
@@ -0,0 +1,229 @@
...
+        self.checkButtons = checkButtons;
+        self.showMoreLess = showMoreLess;
+        self.updateListAll = updateListAll;
+        self.updateListReceived = updateListReceived;
+        self.updateListGave = updateListGave;
+        self.updateListChallenges = updateListChallenges;
+        self.showList = showList;
+        self.getTemplate = getTemplate;
+
+        this.$onInit = function() {
```
Comment (Medds) [inline src/app/components/kudos-user-history/kudos-user-history.component.js]: you could use `self` here as well.

File: src/app/components/kudos-user-history/kudos-user-history.component.js [hunk 22]
```diff
@@ -0,0 +1,229 @@
...
+            self.currentUser = this.user;
```
Comment (Medds) [inline src/app/components/kudos-user-history/kudos-user-history.component.js]: you could use self here as well.

File: src/app/components/kudos-user-history/kudos-user-history.component.js [hunk 23]
```diff
@@ -0,0 +1,229 @@
...
+            if (this.page == true){
```
Comment (Medds) [inline src/app/components/kudos-user-history/kudos-user-history.component.js]: you could use self here as well.

File: src/app/components/kudos-user-history/kudos-user-history.component.js [hunk 24]
```diff
@@ -0,0 +1,229 @@
...
+                updateListAll(self.currentUser.email, true);
+            }
+            self.modalIndex = split(this.user.$$hashKey, ":")[1];
```
Comment (Medds) [inline src/app/components/kudos-user-history/kudos-user-history.component.js]: You could use `self` here as well.

File: src/app/components/kudos-user-history/kudos-user-history.component.js [hunk 25]
```diff
@@ -0,0 +1,229 @@
...
+        }
+
+        function changeRadioValue(value) {
+            if (self.historyRadioBox != value) {
+                self.historyRadioBox = value;
+            }
+        }
+
+        function changeView(value) {
+            if (value == 'all') {
```
Comment (Medds) [inline src/app/components/kudos-user-history/kudos-user-history.component.js]: how about a `switch` here?

File: src/app/components/kudos-user-history/kudos-user-history.component.js [hunk 26]
```diff
@@ -0,0 +1,229 @@
...
+                self.challengeOperations=true;
+            }
+
+            self.showLessButton = false;
+            self.transactionEndingIndex = 6;
+            self.historyRadioBox = 'wonLost';
+        }
+
+        function getTemplate(page) {
+            if (this.page == true){
```
Comment (Medds) [inline src/app/components/kudos-user-history/kudos-user-history.component.js]: You could use `self` here as well.

File: src/app/components/kudos-user-history/kudos-user-history.component.js [hunk 27]
```diff
@@ -0,0 +1,229 @@
...
+
+    };
+
+    UserHistoryController.$inject = ['Resources', 'UserHistoryService', '$timeout'];
+
+    angular.module('myApp.components.userHistory', [])
+        .component('kudosUserHistory', {
+            template: '<ng-include src="self.getTemplate()"/>' ,
+            bindings: {
+                user: '=',
```
Comment (Medds) [inline src/app/components/kudos-user-history/kudos-user-history.component.js]: Are you updating `self.user` anywhere? Do you need this two-way-binding? One way bindings were introduced to components, since usually components only need read-access to any data that is provided into their isolate scope and thus, there is no need to watch for the changes on the variables on the parent controller. This makes it faster and less prone to errors.

File: src/app/components/kudos-user-history/kudos-user-history.component.js [hunk 28]
```diff
@@ -0,0 +1,229 @@
...
+                page: '<'
+            },
+            controller: 'UserHistoryController',
+            controllerAs: 'self'
```
Comment (Medds) [inline src/app/components/kudos-user-history/kudos-user-history.component.js]: An angular.component has controllerAs set to $ctrl by default without declaring anything. It's common practice to use this, since it leads to less code and less confusion. If you're not using the default accessor, you could set it to something semantically correct, which would be `history` or `userHistory` for this instance, so it would be easily understandable.
TL;DR - best to not declare anything and use the default $ctrl, otherwise - semantify.

File: src/app/components/kudos-user-history/kudos-user-history.component.js [hunk 29]
```diff
@@ -0,0 +1,229 @@
...
+        })
+        .controller('UserHistoryController', UserHistoryController)
```
Comment (Medds) [inline src/app/components/kudos-user-history/kudos-user-history.component.js]: Single responsibility principle - one component per file. I'd have no problem if the controller was declared via function, but here you are declaring a new component.

File: src/app/components/kudos-user-history/kudos-user-history.service.js [hunk 30]
```diff
@@ -0,0 +1,161 @@
...
+            getGave : getGaveUserTransactions,
+            setGave : setGaveUserTransactions,
+            getChallenges : getChallengesUserTransactions,
+            setChallenges : setChallengesUserTransactions
+        };
+        return service;
+
+        function getAllTransactions(requestData) {
+            if (requestData){
+                requestData = $httpParamSerializer({
```
Comment (Medds) [inline src/app/components/kudos-user-history/kudos-user-history.service.js]: Do you really need `$httpParamSerializer` here? I was under the influence, that` as long as you provide the correct `Content-Type` header, everything is serialized automatically?

File: src/app/components/kudos-wisdom-wall-ideas/kudos-wisdom-wall-ideas.component.js [hunk 31]
```diff
@@ -1,8 +1,10 @@
 (function () {
     var WisdomWallIdeasController = function ($scope, WisdomWallIdeasService){
+        $scope.showLoader = true;
```
Comment (Medds) [inline src/app/components/kudos-wisdom-wall-ideas/kudos-wisdom-wall-ideas.component.js]: Perfect candidate to remove the `angular.controller` and just use the controller as a function.

File: src/app/login/login.controller.js [hunk 32]
```diff
@@ -101,9 +112,22 @@
...
+                return false;
+            }
+        }
+
     }
 
-    LoginController.$inject = ['$scope', '$cookies', '$base64', '$httpParamSerializer', 'LoginService'];
+    LoginController.$inject = ['$scope', '$cookies', '$base64', '$httpParamSerializer', 'LoginService', '$location'];
 
     angular
```
Comment (Medds) [inline src/app/login/login.controller.js]: Single responsibility principle - one component per file. I'd have no problem if the controller was declared via function, but here you are declaring a new component.

File: src/app/login/login.service.js [hunk 33]
```diff
@@ -31,8 +31,10 @@
 
         function CheckUser() {
             return authBackend.check().then(function (val) {
-                userLoggedIn(val.data.logged) ? changeViewToProfile() : changeViewToLogin();
+                //userLoggedIn(val.data.logged) ? changeViewToProfile() : changeViewToLogin();
```
Comment (Medds) [inline src/app/login/login.service.js]: No commented code in release.

File: src/app/profile/profile.controller.js [hunk 34]
```diff
@@ -12,12 +14,11 @@
 
         $scope.usersCollection = [];
         $scope.topReceivers = [];
-        $scope.showLoader = true;
         $scope.buttonDisabled = true;
         $scope.receivedAcorns = false;
         $scope.sentAcorns = true;
         $scope.giveAcorns = false;
-        $scope.received = true;
+        $scope.new = true;
```
Comment (Medds) [inline src/app/profile/profile.controller.js]: What does new mean? Variables should define what they are, this is not very definitive.

File: src/app/services/Challenges.js [hunk 35]
```diff
@@ -44,10 +46,10 @@ function Challenges($http, SERVER) {
         })
     }
 
-    function getGivenChallenges(requestData) {
+    function getCompletedChallenges() {
         return $http({
             method: 'GET',
-            url: SERVER.ip + "/challenges/createdByStatus?" + requestData,
+            url: SERVER.ip + "/challenges/completedChallenges",
             withCredentials: true
```
Comment (Medds) [inline src/app/services/Challenges.js]: Could you not use `$httpProvider.defaults` in some sort of config to set `withCredentials` and, possibly, the domain of the `url` by default for the whole app, so you wouldn't need to do it for EVERY request?

File: src/test/user-history.spec.js [hunk 36]
```diff
@@ -0,0 +1,38 @@
...
+        httpParamSerializer = $httpParamSerializer;
+        $httpBackend = _$httpBackend_;
+        $httpBackend.when('GET', 'app/translations/locale-en.json').respond();
+        $httpBackend.when('POST', 'http://localhost:8080/user/list').respond();
+    }));
+
+    describe('Controller', function() {
+
+        it('', function () {
+
```
Comment (Medds) [inline src/test/user-history.spec.js]: Hmm...

**Comment by Medds** [comment]
Kind of tired for today - will try to look at this at work at some point (it pains me to code review on such a small screen at home). If you have questions or want basic pointers - catch me at work. Also, if you want, we can have a quick intro into Angular (I suck at Java and Javascript) principles at some point, so it's, maybe, easier to write code/tests then.
First note - https://github.com/johnpapa/angular-styleguide/blob/master/a1/README.md - this guy is a LEGEND in the angular world.

**Comment by Medds** [comment]
@vytautassugintas -  https://docs.angularjs.org/api/ngMock/service/$httpBackend#!
This is how you mock `$http`. You can either test whether the request is sent, or you can mock the response - you choose. You don't need to test whether `$http` actually works - it does. What you would want to check is that you are using the correct data and that the response is returned correctly.
